# Problem 1: (15% for CSC 446, 10% for CSC 546)

Suppose a random variable $X$ has the density function as shown the following figure.


![Figure 1](fig1_assn2.png)

1. Write the density function $f(x)$ for this random variable.

Ans:
\begin{cases}
0.7, & 0\le x \le 1 \\ 
0.2, & 1< x \le c \\
0, & \text{otherwise}
\end{cases}

2. What must be the value of $c$?

$0.7$ + $0.2(c-1)$ = $1$

$0.2c$ - $0.2$ = $0.3$

$0.2c$ = $0.5$

Ans: $c = 2.5$

3. Write the CDF $F(x)$ of this random variable.

Ans: 
\begin{cases}
0.7x, & 0\le x<1 \\
0.2x + 0.5, &1\le x\lt2.5 \\ 
1, &x\ge2.5 \\
0, &x \lt 0
\end{cases}

4. What is $E(X)$?

$E(X)$ = $0.35(1^2 - 0^2)$ + $0.1(2.5^2 - 1^2)$

$E(X)$ = $0.35$ + $0.1(5.25)$ = $0.35$ + $0.525$

Ans: $E(X) = 0.875$


5. What is the variance of $X$?

$V(X) = E(X^2)$ - $E(X)^2$ 

$E(X^2) = \frac{0.7}{3}(1^3-0^3) + \frac{0.2}{3}(2.5^3-1^3)$

$E(X^2) = \frac{0.7}{3} + \frac{39}{40} = \frac{29}{24}$

$V(X) = \frac{29}{24} - 0.875^2$

Ans: $V(X) = 0.4427$






# End of Problem 1


# Problem 2 (20% for CSC 446, 10% for CSC 546)

A terminal at an airport has one queue to check-in for the Terminal $1$ passengers. After the
check-in, passengers join along with Terminal $2$ passengers to a Security-Check
as shown in the following figure. 


![Figure 1](fig2_assn2.png)

The arrivals are Poisson distributed with rates $\lambda_1$ and $\lambda_2$ for Terminal $1$ and $2$ passengers, respectively. The service times are exponentially distributed with service rates $\mu_1, \mu_S$ for terminal $1$ check-in and terminal $2$ security-check, respectively.

The average arrival rates are $10$ customers/hour for Terminal $1$ passengers and $20$ customers/hour for Terminal $2$ passengers. The average service times at the Terminal $1$ check-in is $2$ minutes per customer. The security-check maintains a server utilization of $75\%$ all the time.

Answer the following questions with mathematical analysis:

(1) What is the average service time at Security-Check?

Ans: $\rho$ = $\lambda_1 + \lambda_2 \over \mu_s$

$\mu_s$ = $\lambda_1 + \lambda_2 \over \rho$ = $\frac{1}{3}+\frac{1}{6} \over 0.75$ = $\frac{2}{3}$ customers per minutes

average number of passenger per minutes = $\frac{3}{2}$ = $1.5$ minutes per customer 

The average service time at Security-Check is 1.5 minutes per customer



(2) What is the the average time spent by the passengers (including both terminal $1$ passengers and terminal $2$ passengers) at Security-Check?

Ans: $w_2$ per minutes = $\frac{1}{\mu_2-(\lambda_1+\lambda_2)}$ = $\frac{1}{\frac{2}{3}-(\frac{1}{3} + \frac{1}{6})}$ = $6$ minutes

6 minutes by terminal $1$ and terminal $2$ 


(3) What is the average total time spent in the system by Terminal $1$ passengers(Note that Terminal $1$ passengers need to pass Terminal $1$ and Terminal $2$)?

$w_t$ = $w_1$ + $w_2$

$w_1$ = $\frac{1}{\mu_1 - \lambda_1}$ = $\frac{1}{\frac{1}{2} - \frac{1}{6}}$ = $3$ minutes

$w_t$ = $3$ + $6$ = $9$ minutes

Ans: The average total time spent in the system by terminal $1$ passenger is $9$ minutes


(4) What is the average total number of Terminal $1$ passengers in the system (Note that Terminal $1$ passengers need to pass Terminal $1$ and Terminal $2$)?

$L_t$ = $\lambda_1 w_t$ = $\frac{1}{6}$ $9$ = $1.5$ passengers

Ans: $1.5$ passengers

(5) What is the average time spent by Terminal $2$ passengers in the system?

$w_t$ = $w_q + w_s$

$w_q$ = $\frac{\frac{1}{2}}{\frac{2}{3}(\frac{2}{3}-\frac{1}{2})}$ = $4.5$ minutes

$w_s$ = $1.5$ minutes

$w_t$ = $1.5$ + $4.5$ = $6$ minutes

Ans: $6$ minutes


# End of Problem 2


# Problem 3 (50% for CSC 446, 50% for CSC 546)
Simulate the system in Problem 2 by extending Hritik Jaiswal's code as below.

In [2]:
'''
DO NOT TOUCH THIS CODE!!
Owner : Hritik Jaiswal
Topic : To simulate a Single Server Queuing System (One-operator Barbershop problem) using Python
Subject : Modeling and simulation
'''

import random 

# Seed 
random.seed(10)

# No. of Customer
size = 10

# Series of customer
customer = [i for i in range(1,size+1)]

# Inter Arrival Time 
inter_arrival_time = [random.randrange(1,10) for i in range(size)]

# Service Time
service_time = [random.randrange(1,10) for i in range(size)]

print(len(inter_arrival_time),len(service_time))

# Calculate arrival time
arrival_time = [0 for i in range(size)]

# initial
arrival_time[0] = inter_arrival_time[0]

for i in range(1,size):
  arrival_time[i] = inter_arrival_time[i]+arrival_time[i-1]
 

Time_Service_Begin = [0 for i in range(size)]
Time_Customer_Waiting_in_Queue = [0 for i in range(size)]
Time_Service_Ends = [0 for i in range(size)]
Time_Customer_Spend_in_System = [0 for i in range(size)]
System_idle = [0 for i in range(size)]

Time_Service_Begin[0] = arrival_time[0]
Time_Service_Ends[0] = service_time[0]
Time_Customer_Spend_in_System[0] = service_time[0]
Time_Service_Ends[0] = Time_Service_Begin[0] + service_time[0]

for i in range(1,size):
  # Time Service Begin 
  Time_Service_Begin[i] = max(arrival_time[i],Time_Service_Ends[i-1])

  # Time customer waiting in queue   
  Time_Customer_Waiting_in_Queue[i] = Time_Service_Begin[i]-arrival_time[i]

  # Time service ends
  Time_Service_Ends[i] = Time_Service_Begin[i] + service_time[i]  

  # Time Customer Spend in the system
  Time_Customer_Spend_in_System[i] = Time_Service_Ends[i] - arrival_time[i]
  

  # Time when system remains idle
  if (arrival_time[i]>Time_Service_Ends[i-1]):
    System_idle[i] = arrival_time[i]-Time_Service_Ends[i-1]
  else:
    System_idle[i] = 0 
   

from prettytable import PrettyTable

x = PrettyTable()

column_names = ['Customer','IAT','AT','ST','TSB','TCWQ','TSE','TCSS','System Idle']
data = [customer,inter_arrival_time,arrival_time,service_time, Time_Service_Begin, Time_Customer_Waiting_in_Queue, Time_Service_Ends, Time_Customer_Spend_in_System, System_idle]

length = len(column_names)

for i in range(length):
  x.add_column(column_names[i],data[i])
  
print(x)


10 10
+----------+-----+----+----+-----+------+-----+------+-------------+
| Customer | IAT | AT | ST | TSB | TCWQ | TSE | TCSS | System Idle |
+----------+-----+----+----+-----+------+-----+------+-------------+
|    1     |  1  | 1  | 9  |  1  |  0   |  10 |  9   |      0      |
|    2     |  7  | 8  | 8  |  10 |  2   |  18 |  10  |      0      |
|    3     |  8  | 16 | 6  |  18 |  2   |  24 |  8   |      0      |
|    4     |  1  | 17 | 2  |  24 |  7   |  26 |  9   |      0      |
|    5     |  4  | 21 | 4  |  26 |  5   |  30 |  9   |      0      |
|    6     |  8  | 29 | 6  |  30 |  1   |  36 |  7   |      0      |
|    7     |  8  | 37 | 1  |  37 |  0   |  38 |  1   |      1      |
|    8     |  5  | 42 | 7  |  42 |  0   |  49 |  7   |      4      |
|    9     |  3  | 45 | 3  |  49 |  4   |  52 |  7   |      0      |
|    10    |  1  | 46 | 6  |  52 |  6   |  58 |  12  |      0      |
+----------+-----+----+----+-----+------+-----+------+-------------+


Your extended code should be written below: 

In [1]:
import random
import pandas as pd
import numpy as np

# Seed 
random.seed(30)

# No. of Customer
size = 100000

# Series of customer
customer_one = [i for i in range(1,size+1)]
customer_two = [i for i in range(1,size+1)]

# Inter Arrival Time 
l_1 = 1/6
l_2 = 1/3
inter_arrival_time_one = [random.expovariate(l_1) for i in range(1, size+1)]
inter_arrival_time_two = [random.expovariate(l_2) for i in range(1, size+1)]

# Service Time
mu_1 = 1/2
mu_2 = 2/3
service_time_one = [random.expovariate(mu_1) for i in range(size)]
service_time_two = [random.expovariate(mu_2) for i in range(size)]


#print(len(inter_arrival_time_one))
'''print(f'customer_1: \n{customer_one}')
print(f'\ncustomer_2: \n{customer_two}')'''
print()

In [3]:
# Calculate arrival time
arrival_time_one = [0]*size 
arrival_time_two = [0]*size

# initial
arrival_time_one[0] = inter_arrival_time_one[0]
arrival_time_two[0] = inter_arrival_time_two[0]

# calculate arrival time for terminal 1 and 2 passengers
for i in range(1,size):
    arrival_time_one[i] = inter_arrival_time_one[i]+arrival_time_one[i-1]
    arrival_time_two[i] = inter_arrival_time_two[i]+arrival_time_two[i-1]

# initialize variables
Time_Service_Begin_one = [0]*size
#Time_Service_Begin_two = [0]*size 

Time_Customer_Waiting_in_Queue_one = [0 for i in range(size)]
#Time_Customer_Waiting_in_Queue_two = [0 for i in range(size)]

Time_Service_Ends_one = [0 for i in range(size)]
#Time_Service_Ends_two = [0 for i in range(size)]

Time_Customer_Spend_in_System_one = [0 for i in range(size)]
#Time_Customer_Spend_in_System_two = [0 for i in range(size)]

System_idle_one = [0 for i in range(size)]
#System_idle_two = [0 for i in range(size)]

# set the first value in the list
Time_Service_Begin_one[0] = arrival_time_one[0]
#Time_Service_Begin_two[0] = arrival_time_two[0]

Time_Service_Ends_one[0] = Time_Service_Begin_one[0] + service_time_one[0]
#Time_Service_Ends_two[0] = Time_Service_Begin_two[0] + service_time_two[0]

Time_Customer_Spend_in_System_one[0] = service_time_one[0]
#Time_Customer_Spend_in_System_two[0] = service_time_two[0]

'''print('arrival time 1')
print(arrival_time_one)
print(f'\nInter_arrival_time_one: {inter_arrival_time_one}\n')


print('\narrival time 2')
print(arrival_time_two)
print(f'\nInter_arrival_time_two: {inter_arrival_time_two}\n')'''
print()

In [4]:
for i in range(1,size):
    # Time Service Begin 
    Time_Service_Begin_one[i] = max(arrival_time_one[i],Time_Service_Ends_one[i-1])
    #Time_Service_Begin_two[i] = max(arrival_time_two[i],Time_Service_Ends_two[i-1])
    
    # Time customer waiting in queue   
    Time_Customer_Waiting_in_Queue_one[i] = Time_Service_Begin_one[i]-arrival_time_one[i]
    #Time_Customer_Waiting_in_Queue_two[i] = Time_Service_Begin_two[i]-arrival_time_two[i]

    # Time service ends
    Time_Service_Ends_one[i] = Time_Service_Begin_one[i] + service_time_one[i]  
    #Time_Service_Ends_two[i] = Time_Service_Begin_two[i] + service_time_two[i]  

    # Time Customer Spend in the system
    Time_Customer_Spend_in_System_one[i] = Time_Service_Ends_one[i] - arrival_time_one[i]
    #Time_Customer_Spend_in_System_two[i] = Time_Service_Ends_two[i] - arrival_time_two[i]

    # Time when system remains idle for terminal 1
    if (arrival_time_one[i]>Time_Service_Ends_one[i-1]):
        System_idle_one[i] = arrival_time_one[i]-Time_Service_Ends_one[i-1]
    else:
        System_idle_one[i] = 0 
        
    # Time when system remains idle for terminal 2
    '''if (arrival_time_two[i]>Time_Service_Ends_two[i-1]):
        System_idle_two[i] = arrival_time_two[i]-Time_Service_Ends_two[i-1]
    else:
        System_idle_two[i] = 0 '''

'''print('arrival time 1')
print(arrival_time_one)
print('\narrival time 2')
print(arrival_time_two)'''
print()


In [5]:
column_names = ['Customer','AT','ST','TSB','TCWQ','TSE','TCSS','System Idle', 'AT security check']

length = len(column_names)

# store all customers based on their arrival time and terminal key
key = ['Terminal'] + column_names
customer_queue = []

# store terminal 1 and 2
for i in range(len(arrival_time_one)):
    t_1_cust = [1, customer_one[i], arrival_time_one[i], service_time_one[i], Time_Service_Begin_one[i],
               Time_Customer_Waiting_in_Queue_one[i], Time_Service_Ends_one[i], Time_Customer_Spend_in_System_one[i], 
                System_idle_one[i], Time_Service_Ends_one[i]]
    customer_queue.append(t_1_cust)
    
'''for i in range(len(arrival_time_one)):
    t_2_cust = [2, customer_two[i], arrival_time_two[i], service_time_two[i], Time_Service_Begin_two[i], 
               Time_Customer_Waiting_in_Queue_two[i], Time_Service_Ends_two[i], Time_Customer_Spend_in_System_two[i], 
                System_idle_two[i], arrival_time_two[i]]
    customer_queue.append(t_2_cust)'''

df = pd.DataFrame(customer_queue, columns=key)

# sort data from the top 30 arrival time in terminal 1 and 2
df.sort_values(by = ['AT security check'], ascending = True, inplace = True)
top = df.head(size)

#top
print()

In [6]:
# count frequency of terminal 1 and 2
freq_one = len(np.where(np.array(top['Terminal'])==1)[0])
#freq_two = len(np.where(np.array(top['Terminal'])==2)[0])
        
print(f'\nTerminal 1: {freq_one}')
#print(f'Terminal 2: {freq_two}')



Terminal 1: 100000


In [7]:
# simulate security check

terminal = np.array(list(top['Terminal'])+list([2]*size))
old_arrival = np.array(list(top['AT'])+list(arrival_time_two))
security_arrival = np.array(list(top['AT security check'])+list(arrival_time_two))
service_t = np.array(list(top['ST'])+list(service_time_two))

table = []
for i in range(2*size):
    table.append([old_arrival[i], security_arrival[i], service_t[i], terminal[i]])

table.sort(key = lambda x: x[1])


tsb = [0]*2*size
twq = [0]*2*size
tse = [0]*2*size
tss = [0]*2*size
sys_idle = [0]*2*size

tsb[0] = table[0][1]
tse[0] = tsb[0] + table[0][2]
tss[0] = table[0][2]


for i in range(1, size):
    # Time Service Begin 
    tsb[i] = max(table[i][1], tse[i-1])
    
    # Time customer waiting in queue   
    twq[i] = tsb[i] - table[i][1]

    # Time service ends
    tse[i] = tsb[i] + table[i][2]  

    # Time Customer Spend in the system
    tss[i] = tse[i] - table[i][1]

    # Time when system remains idle
    if (table[i][1]>tse[i-1]):
        sys_idle[i] = table[i][1]-tse[i-1]
    else:
        sys_idle[i] = 0 
        
'''print(f'TSE for terminal 1 after check in: \n{tse}')

print(f'\nArrival time for terminal 2: \n{arrival_time_two}')'''
print()

In [8]:
# create 2d list of terminal 1 customer after check in
key = ['Terminal', 'AT','ST','TSB','TCWQ','TSE','TCSS','System Idle', 'AT security check']
cust_tot = []

for i in range(2*size):
    cust_tot.append([table[i][3], table[i][0], table[i][2], tsb[i], twq[i], tse[i], tss[i], sys_idle[i], table[i][1]])
    

cust_tot.sort(key = lambda x: x[8])
        
#for row in cust_tot:
#    print(row)

In [9]:
top = pd.DataFrame(cust_tot, columns=key)
top = top.head(size)
#top
print()

In [10]:
# output simulation result
avg_service = top['ST'].sum()/len(top['ST'])
print(f'avg service time: {avg_service}\n')

# sum(tcss) / len(tcss)
avg_sys = top['TCSS'].sum()/len(top['TCSS'])
print(f'average time in security check: {avg_sys}\n')

# 
terminal_list = np.array(top['Terminal'])
t1_index = np.where(terminal_list==1)[0]
tot_sys_t1 = np.sum(np.subtract(np.array(top['TSE'])[t1_index], np.array(top['AT'])[t1_index])) 
print(f'T1 avg total time in the system: {tot_sys_t1 / len(t1_index)}\n')

# t1 total system time / last tse
tot_passenger = tot_sys_t1 / np.array(top['TSE'])[-1]
print(f'Total number T1 in system: {tot_passenger} passengers per min\n')

# t2 total system time / # t2 
t2_index = np.where(terminal_list==2)[0]
sys_t_t2 = np.sum(np.array(top['TCSS'])[t2_index]) / len(t2_index)
print(f'Average time in system T2: {sys_t_t2}')

avg service time: 1.667859703987526

average time in security check: 10.490304379307124

T1 avg total time in the system: 14.199287776567612

Total number T1 in system: 2.387276190130442 passengers per min

Average time in system T2: 10.131522399573395


Simulate $30$ customers in total. Compare the theoretical results calculated in Problem 2 and the simulation results obtained in Problem 3. Use a table to show the comparison results.


Ans: 


|  Metric                                            | Analytical result   | Simulation Result | 
|:---------------------------------------------------| :-----------------: | :--------------:  |                
|  avg. service time at Security-Check               |     1.5 minutes     |   1.4191 minutes  |  
|  avg. time spent by the passengers                 |     6 minutes       |   3.0278 minutes  |  
|  avg. total time in the system (T1 passengers)     |     9 minutes       |   5.1448 minutes  |
|  avg. total number of T1 passengers in the system  |   1.5 passengers    |  0.7937 passenger |
|  avg. time spent by T2 passengers in the system    |     6 minutes       |   2.9765 minutes  |  



Increasing the number of simulated customers to $100,000$. Run your simulation five times, each run with a different random seed. Record the simulation result of each run and take the average over the five runs as the final simulation result. (Important: to avoid bloating your output screen, you should disable the output table when you simulate $100,000$ customers.)  

Ans: 


|  Metric                                            | Run 1  | Run 2 | Run 3 | Run 4 | Run 5 | Final Result | 
|:---------------------------------------------------| :----: | :---: |:----: | :--:--|:----: | :---------:  |             
|  avg. service time at Security-Check               |1.667859|1.67339|1.66706|1.66520|1.66790|  1.6682818   |  
|  avg. time spent by the passengers                 |10.49030|10.6804|10.1444|10.3189|10.6226|  10.4372456  |
|  avg. total time in the system (T1 passengers)     |14.19928|14.4441|13.9401|14.0860|14.4842|  14.230736   |
|  avg. total number of T1 passengers in the system  |2.387271|2.40399|2.30768|2.34805|2.43301|  2.3760002   |
|  avg. time spent by T2 passengers in the system    |10.13152|10.3158|9.74900|9.94559|10.2047|  10.069322   |





# End of Problem 3


# Problem 4 (15% for CSC 446, 10% for CSC 546)

Let $W$ and $W_Q$ denote the total waiting time of a packet in the system and the queueing time of the packet, respectively. 

1. Is $W$ the same in an M/M/1 model having arrivals at rate λ and service at rate 2µ as it is in a two-server M/M/2 model with arrivals at rate λ and with each server at rate µ? Prove your claim. 

$\rho = \frac{\lambda}{2\mu}$

$P_0 = [1 + 2\rho + \frac{2\rho^2}{1-\rho}]^{-1} = [\frac{1 + \rho}{1 - \rho}]^{-1} = \frac{1 - \rho}{1 + \rho}$

$L = 2\rho + \frac{(2\rho)^{3}(1-\rho)}{4(1-\rho)^{2}(1+\rho)} = \frac{2\rho^3}{(1-\rho^2)}$

$L = \frac{4\mu\lambda}{(2\mu-\lambda)(2\mu+\lambda)}$

$W_2 = \frac{L}{\lambda} = \frac{4\mu}{(2\mu-\lambda)(2\mu+\lambda)}$

Since $W_1 = \frac{1}{2\mu-\lambda} \ne W_2$ 

Ans: $W(M/M/2) \ne W(M/M/1)$

2. Is $W_Q$ the same in an M/M/1 model having arrivals at rate λ and service at rate 2µ as it is in a two-server M/M/2 model with arrivals at rate λ and with each server at rate µ? Prove your claim.

For $W_Q$ in M/M/1

$W_Q = \frac{\lambda}{2\mu(2\mu-\lambda)}$

For $W_Q$ in M/M/2

$W_Q = W - \frac{1}{\mu} = \frac{\lambda^2}{\mu(2\mu-\lambda)(2\mu+\lambda)}$

Compare $\frac{\lambda}{2\mu(2\mu-\lambda)}$ with $\frac{\lambda^2}{\mu(2\mu-\lambda)(2\mu+\lambda)}$

And compare $\frac{1}{2}$ with $\frac{\lambda}{2\mu+\lambda}$

$\rho = \frac{\lambda}{2\mu} \lt 1$ 

Therefore $\frac{1}{2} \gt \frac{\lambda}{2\mu+\lambda}$ and $\frac{\lambda}{2\mu(2\mu-\lambda)} \gt \frac{\lambda^2}{\mu(2\mu-\lambda)(2\mu+\lambda)}$

Ans: $W_Q(M/M/1) \ne W_Q(M/M/2)$


# End of Problem 4


# Problem 5 (CSC 546 Only, 20% )

Patients arrive for a physical examination according to a Poisson process at the rate 1 per hour. The physical examination requires three stages, each one independently exponentially distributed with a mean service time of 15 minutes. A patient must go through all the three stages before the next patient is admitted to the treatment facility. 

1. What is the mean and variance of the total examination time? 

Ans:


2. Is the total examination time an exponential distribution? 

Ans:


3. Compute the average number of delayed patients, $L_Q$ for this system. 

Ans:

4. Compute the total mean time a customer spends in the system.

Ans:



# End of Problem 5